datafusion/common/src/cse.rs

```diff
@@ -0,0 +1,788 @@
+            let (count, conditional_count) =
+                self.node_stats.entry(node_id).or_insert((0, 0));
+            if self.conditional {
+                *conditional_count += 1;
+            } else {
+                *count += 1;
+            }
+            if *count > 1 || (*count == 1 && *conditional_count > 0) {
+                self.found_common = true;
+            }
+        }
+        self.visit_stack
+            .push(VisitRecord::NodeItem(node_id, is_valid));
+        self.up_index += 1;
+
+        Ok(TreeNodeRecursion::Continue)
+    }
+}
+
+/// Rewrite a [`TreeNode`] tree by replacing detected common subtrees with the
+/// corresponding temporary [`TreeNode`], that column contains the evaluate result of
+/// replaced [`TreeNode`] tree.
+struct CommonSubTreeNodeRewriter<'a, 'n, N, C: SubTreeNodeEliminatorController<Node = N>>
+{
+    /// statistics of [`TreeNode`]s
+    node_stats: &'a NodeStats<'n, N>,
+
+    /// cache to speed up second traversal
+    id_array: &'a IdArray<'n, N>,
+
+    /// common [`TreeNode`]s, that are replaced during the second traversal, are collected
+    /// to this map
+    common_nodes: &'a mut CommonNodes<'n, N>,
+
+    // preorder index, starts from 0.
+    down_index: usize,
+
+    controller: &'a mut C,
+}
+
+impl<N: TreeNode + Eq, C: SubTreeNodeEliminatorController<Node = N>> TreeNodeRewriter
+    for CommonSubTreeNodeRewriter<'_, '_, N, C>
+{
+    type Node = N;
+
+    fn f_down(&mut self, node: Self::Node) -> Result<Transformed<Self::Node>> {
+        self.controller.rewrite_f_down(&node);
+
+        let (up_index, node_id) = self.id_array[self.down_index];
+        self.down_index += 1;
+
+        // Handle nodes with identifiers only
+        if let Some(node_id) = node_id {
+            let (count, conditional_count) = self.node_stats.get(&node_id).unwrap();
+            if *count > 1 || *count == 1 && *conditional_count > 0 {
+                // step index to skip all sub-node (which has smaller series number).
+                while self.down_index < self.id_array.len()
+                    && self.id_array[self.down_index].0 < up_index
+                {
+                    self.down_index += 1;
+                }
+
+                let (node, alias) =
+                    self.common_nodes.entry(node_id).or_insert_with(|| {
+                        let node_alias = self.controller.generate_alias();
+                        (node, node_alias)
+                    });
+
+                let rewritten = self.controller.rewrite(node, alias);
+
+                return Ok(Transformed::new(rewritten, true, TreeNodeRecursion::Jump));
+            }
+        }
+
+        Ok(Transformed::no(node))
+    }
+
+    fn f_up(&mut self, node: Self::Node) -> Result<Transformed<Self::Node>> {
+        self.controller.rewrite_f_up(&node);
+
+        Ok(Transformed::no(node))
+    }
+}
+
+pub struct SubTreeNodeEliminator<N, C: SubTreeNodeEliminatorController<Node = N>> {
```

Comment:
Is this the main interface to the CSE code? If so, can we perhaps add some documentation comments explaining what this does / how to use it? 

I am thinking when someone else is looking at this code outside the context of this PR and wants to know what it is used for / does

Reply:
Done in https://github.com/apache/datafusion/pull/13002/commits/1cae61cffb296fd7bd546e6d3f687f85496ab591.